File: 0022/0022_generate_parentheses.go
package main

import "fmt"

func next_permutation(arr []byte) bool {
	n := len(arr) - 1
	k := -1
	for i := n - 1; i >= 0; i-- {
		if arr[i] < arr[i+1] {
			k = i
			break
		}
	}
	// last permutation
	if k == -1 {
		return false
	}
	l := n
	for i := n; i > k; i-- {
		if arr[k] < arr[i] {
			l = i
			break
		}
	}
	arr[k], arr[l] = arr[l], arr[k]
	i := k + 1
	j := n
	for i < j {
		arr[i], arr[j] = arr[j], arr[i]
		i++
		j--
	}
	return true
}

func is_wellformed(str []byte) bool {
	c := 0
	for i := 0; i < len(str); i++ {
		if str[i] == '(' {
			c++
		} else if str[i] == ')' {
			c--
			if c < 0 {
				return false
			}
		}
	}
	return c == 0
}

func generateParenthesis(n int) []string {
	var arr []byte
	var result []string
	for i := 0; i < n; i++ {
		arr = append(arr, '(')
	}
	for i := 0; i < n; i++ {
		arr = append(arr, ')')
	}
	// this is must be well-formed
	result = append(result, string(arr))
	for next_permutation(arr) {
		if is_wellformed(arr) {
			result = append(result, string(arr))
		}
	}
	return result
}

func main() {
	result := generateParenthesis(1)
	for i := 0; i < len(result); i++ {
		fmt.Println(result[i])
	}
}

File: 0022/0022_generate_parentheses.rs
struct Solution;

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut results: Vec<String> = vec![];
        let mut buf: Vec<char> = vec![];
        for _ in 0..n {
            buf.push('(');
        }
        for _ in 0..n {
            buf.push(')');
        }
        // this is must be well-formed
        results.push(buf.iter().collect());
        while Self::next_permutation(&mut buf) {
            if Self::is_wellformed(&buf) {
                results.push(buf.iter().collect());
            }
        }
        results
    }

    fn next_permutation(buf: &mut Vec<char>) -> bool {
        let n = buf.len() - 1;
        let mut k2: Option<usize> = None;
        for i in (0..=n-1).rev() {
            if buf[i] < buf[i + 1] {
                k2 = Some(i);
                break;
            }
        }
        // last permutation
        if k2.is_none() {
            return false;
        }
        let k = k2.unwrap();
        let mut l = n;
        for i in (k + 1..=n).rev() {
            if buf[k] < buf[i] {
                l = i;
                break;
            }
        }
        buf.swap(k, l);
        let mut i = k + 1;
        let mut j = n;
        while i < j {
            buf.swap(i, j);
            i += 1;
            j -= 1;
        }
        true
    }

    fn is_wellformed(buf: &Vec<char>) -> bool {
        let mut c = 0;
        let mut e = false;
        buf.iter().for_each(|x| {
            if x == &'(' {
                c += 1;
            } else {
                c -= 1;
                if c < 0 {
                    e = true
                }
            }
        });
        !e && c == 0
    }
}

fn main() {
    assert_eq!(vec!["()"], Solution::generate_parenthesis(1));
    assert_eq!(vec!["(())", "()()"], Solution::generate_parenthesis(2));
    assert_eq!(
        vec!["((()))", "(()())", "(())()", "()(())", "()()()"],
        Solution::generate_parenthesis(3)
    );
}
